tests/unit/threaded_deque_test.c

```diff
@@ -0,0 +1,426 @@
+#include <test.h>
+
+#include <alloc.h>
+#include <mutex.h>
+#include <threaded_deque.h>
+
+/* Memory illustration legend:          *
+ *      | : memory bounds               *
+ *      > : left                        *
+ *      < : right                       *
+ *      ^ : left + right (empty)        *
+ *      v : left + right (at capacity)  *
+ *      x : used memory                 *
+ *      - : unused memory               */
+
+static void test_push_pop(void)
+{
+    // Initialised with DEFAULT_CAPACITY = 16
+    ThreadedDeque *deque = ThreadedDequeNew(0, free);
+    // |^---------------|
+    ThreadedDequePushLeft(deque, xstrdup("1"));
+    // |<-------------->|
+    ThreadedDequePushRight(deque, xstrdup("2"));
+    // |x<------------->|
```

Comment:
This doesn't look like a result of `PushRight` to me.

Reply:
Whoops. Thanks for noticing!

Reply:
Wait, yes it is! You push to the right index.

Reply:
It looks confusing, I know.

Reply:
So you push to the right and the item is on the left? :confused:

Reply:
It pushes to the _right end_ of the queue. This means that since the array has wrapped around by first pushing to the left, it looks as if you are pushing on the left, I guess?

If you had only do pushes right, it would look like this:
```
|><--------------|
|>x<-------------|
```
instead of:
```
|<-------------->|
|x<------------->|
```

Reply:
OK, analyzed this with @olehermanse and it makes sense. Maybe the legend should explicitly say that `<` and `>` are positions where the `Pop()` would take its item from not where `Push()` would push its item. That would also make it clear that `<` is actually also `x` (occupied memory).